python/cudf/cudf/tests/test_index.py

```diff
@@ -396,7 +398,18 @@ def test_index_copy_category(name, dtype, deep=True):
 def test_index_copy_deep(idx, deep):
     """Test if deep copy creates a new instance for device data."""
     idx_copy = idx.copy(deep=deep)
-    if not deep:
+
+    if (
+        isinstance(idx, cudf.StringIndex)
+        or not deep
+        or cudf.get_option("copy_on_write")
+    ):
+        # StringColumn is immutable hence, deep copies of a
+        # StringIndex will share the same StringColumn.
+
+        # When `copy_on_write` is turned on, Index objects will
+        # have unique column object but they all point to same
+        # data pointers.
```

Comment:
Historically, allowing access to the underlying data ptr is something that we've talked about cuDF supporting. Having a deep copy return the same ptr where someone can then do something like run some custom code or a numba kernel or something of the likes that mutates that underlying data is then unexpected behavior and a breaking change.

I understand from a convenience perspective and not having to crawl through the codebase and touch everything, but I'm -1 on changing the behavior of how an explicitly asked for deep copy works.

Reply:
@kkraus14 When you say allowing access to the underlying data ptr, do you expect `Column.data_ptr` to be the accessor or `Column.data.ptr` to be the accessor? Currently, we support both, but to avoid mutations passing onto other deep copies I could prevent that from happening via `Column.__cuda_array_interface__` & `Column.data_array_view`. This could be possible with `Column.data_ptr` but wanting to know if `Series._column.data.ptr` is something we also want to support ?

Reply:
I'm not sure what we want to expose as a "public" API or not, but we currently support a `.data` attribute on `Series` objects that returns a `Buffer` which has a `.ptr` attribute which returns the underlying pointer as an integer and exposing the underlying pointers is something we have historically presented as a supported feature.

We could detach in the path of accessing the underlying `Buffer` object, but I'm guessing that would eliminate a lot of the benefits we see here.

Reply:
> We could detach in the path of accessing the underlying `Buffer` object, but I'm guessing that would eliminate a lot of the benefits we see here.

This is just me thinking out loud: How about we detach whenever `Buffer.__cuda_array_interface__` is accessed? Since this is the standard protocol for zero-copy? But don't go into checking for weakrefs and detaching when accessing `Buffer.ptr`? I know that everything we return in `Buffer.__cuda_array_interface__` can be hand created by accessing the respective attributes like `.ptr`, `.size`, etc too but should we guarantee a detached copy while accessing `Buffer.ptr` is accessed, reason I'm asking this is we use this property a lot in the code and that could probably trigger a copy pre-emptively before even handing off to another library as a zero-copy

Reply:
Update: I've added a detaching mechanism when `__cuda_array_interface__` is triggered. @kkraus14 could to take a look at it again and see if there is still a concern?